cmd/hub/mcscontroller/controller.go

```diff
@@ -0,0 +1,169 @@
+/*
+SPDX-License-Identifier: Apache-2.0
+
+Copyright Contributors to the Submariner project.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	mcsv1a1 "sigs.k8s.io/mcs-api/pkg/apis/v1alpha1"
+
+	"github.com/submariner-io/lighthouse/pkg/mcs"
+)
+
+const (
+	serviceExportFinalizer = "lighthouse.submariner.io/service-export-finalizer"
+)
+
+// ServiceExportReconciler reconciles a ServiceExport object
+type ServiceExportReconciler struct {
+	Client client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=multicluster.x-k8s.io,resources=serviceexports,verbs=get;list;watch;update;patch
+// +kubebuilder:rbac:groups=multicluster.x-k8s.io,resources=serviceexports/finalizers,verbs=get;update;delete
+// +kubebuilder:rbac:groups=multicluster.x-k8s.io,resources=serviceexports/status,verbs=get;update
+// +kubebuilder:rbac:groups=multicluster.x-k8s.io,resources=serviceimports,verbs=create;get;list;update;patch;delete
+
+// Reconcile handles an update to a ServiceExport
+// By design, the event that triggered reconciliation is not passed to the reconciler
+// to ensure actions are based on state alone. This approach is referred to as level-based,
+// as opposed to edge-based.
+func (r *ServiceExportReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	log := r.Log.WithValues("serviceexport", req.NamespacedName)
+
+	// @todo extract and log the real (i.e., source) object name, namespace and cluster?
+	log.Info("Reconciling ServiceExport")
+
+	se := mcsv1a1.ServiceExport{}
+	if err := r.Client.Get(ctx, req.NamespacedName, &se); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("No ServiceExport found")
```

Comment:
In what situation this can happen? Isn't it considered an error as well?

Reply:
Will this happen if for some reason we could not handle the deletion even in time while finalizer was active? What happens then? no recon will occur right?